src/main.rs

```diff
@@ -320,35 +324,56 @@ async fn entry(jobserver_client: LazyJobserverClient) -> Result<()> {
     // Initialize UI thread
     let mut uithread = UIThread::new(!opts.no_confirm);
 
-    // Launch target detection
-    let desired_targets = get_desired_targets(&opts.targets);
+    let (install_path, metadata, temp_dir) = block_in_place(|| -> Result<_> {
+        // Compute install directory
+        let (install_path, custom_install_path) = get_install_path(opts.install_path.as_deref());
+        let install_path = install_path.ok_or_else(|| {
+            error!("No viable install path found of specified, try `--install-path`");
+            miette!("No install path found or specified")
+        })?;
+        fs::create_dir_all(&install_path).map_err(BinstallError::Io)?;
+        debug!("Using install path: {}", install_path.display());
+
+        // Load metadata
+        let metadata = if !custom_install_path {
+            debug!("Reading binstall/crates-v1.json");
+            Some(metafiles::binstall_v1::Records::load()?)
+        } else {
+            None
+        };
 
-    // Compute install directory
-    let (install_path, custom_install_path) = get_install_path(opts.install_path.as_deref());
-    let install_path = install_path.ok_or_else(|| {
-        error!("No viable install path found of specified, try `--install-path`");
-        miette!("No install path found or specified")
+        // Create a temporary directory for downloads etc.
+        //
+        // Put all binaries to a temporary directory under `dst` first, catching
+        // some failure modes (e.g., out of space) before touching the existing
+        // binaries. This directory will get cleaned up via RAII.
+        let temp_dir = tempfile::Builder::new()
+            .prefix("cargo-binstall")
+            .tempdir_in(&install_path)
+            .map_err(BinstallError::from)
+            .wrap_err("Creating a temporary directory failed.")?;
+
+        Ok((install_path, metadata, temp_dir))
     })?;
-    fs::create_dir_all(&install_path).map_err(BinstallError::Io)?;
-    debug!("Using install path: {}", install_path.display());
-
-    // Create a temporary directory for downloads etc.
-    //
-    // Put all binaries to a temporary directory under `dst` first, catching
-    // some failure modes (e.g., out of space) before touching the existing
-    // binaries. This directory will get cleaned up via RAII.
-    let temp_dir = tempfile::Builder::new()
-        .prefix("cargo-binstall")
-        .tempdir_in(&install_path)
-        .map_err(BinstallError::from)
-        .wrap_err("Creating a temporary directory failed.")?;
+
+    // Remove installed crates
+    let crate_names = CrateName::dedup(crate_names).filter(|crate_name| {
+        if opts.force {
+            true
+        } else if let Some(records) = &metadata {
+            !records.contains(&crate_name.name)
```

Comment:
Hold on. I think that would skip installing a crate if it's already installed _regardless of version_? We should consider version here, and only skip installing crates if they are the exact same version as is installed, which is the behaviour of `cargo-install` without `--force` (in both upgrade and downgrade directions).

Reply:
Done!

Now it will skip the installation only if `version_req.match(installed_version)`.